runtime/src/accounts_db.rs

```diff
@@ -3335,41 +3354,58 @@ impl AccountsDB {
 
     pub fn get_accounts_delta_hash(&self, slot: Slot) -> Hash {
         let mut scan = Measure::start("scan");
-        let mut accumulator: Vec<HashMap<Pubkey, (u64, Hash)>> = self.scan_account_storage(
-            slot,
-            |loaded_account: LoadedAccount,
-             _store_id: AppendVecId,
-             accum: &mut HashMap<Pubkey, (u64, Hash)>| {
-                accum.insert(
-                    *loaded_account.pubkey(),
-                    (
-                        loaded_account.write_version(),
-                        *loaded_account.loaded_hash(),
-                    ),
-                );
-            },
-        );
+
+        let scan_result: ScanStorageResult<(Pubkey, Hash, u64), DashMapVersionHash> = self
+            .scan_account_storage(
+                slot,
+                |loaded_account: LoadedAccount| {
+                    // Cache only has one version per key, don't need to worry about versioning
+                    Some((*loaded_account.pubkey(), *loaded_account.loaded_hash(), 0))
+                },
+                |accum: &DashMap<Pubkey, (u64, Hash)>, loaded_account: LoadedAccount| {
+                    let write_version = loaded_account.write_version();
+                    let should_insert =
+                        if let Some(existing_entry) = accum.get(loaded_account.pubkey()) {
+                            write_version > existing_entry.value().version()
+                        } else {
+                            true
+                        };
+                    if should_insert {
+                        match accum.entry(*loaded_account.pubkey()) {
+                            Occupied(mut key_entry) => {
+                                if write_version > key_entry.get().version() {
```

Comment:
why are we checking `write_version` twice in this closure? I think `should_insert` can be just removed?

Reply:
the first check `if let Some(existing_entry) = accum.get(loaded_account.pubkey()) {` grabs the read lock on that shard to avoid the write lock in most cases.

The second check `match accum.entry(*loaded_account.pubkey()) {` grabs the write lock to commit the write, and has to check again in case another thread interleaved a write between the read + write.

Added a comment!